[FLINK-24379][Formats] Avro Glue Schema Registry table format

nicusX opened the pull request:
## Purpose of the change

Implement Table API support for Avro with AWS Glue Schema Registry

## Verifying this change

- Added unit tests
- Manually verified by running a Flink application locally against a local Kafka cluster and using GSR.

## Significant changes

- [ ] Dependencies have been added or upgraded
- [ ] Public API has been changed (Public API is any class annotated with `@Public(Evolving)`)
- [ ] Serializers have been changed
- [X] New feature has been introduced
  - If yes, how is this documented? docs

The implementation is mostly based on the [original PR to the main Flink repo](https://github.com/apache/flink/pull/17360) with minor changes. SQL uber-jar has been added and docs fixed and enriched.

nicusX commented:
@dannycranmer please review this PR when you have the chance

z3d1k reviewed (changes requested):
Thank you for this contribution @nicusX!

Please add integration tests to cover usage of the format using SQL.

nicusX commented on flink-formats-aws/flink-sql-avro-glue-schema-registry/pom.xml:
```diff
@@ -0,0 +1,163 @@
+                                    <pattern>com.amazonaws</pattern>
+                                    <shadedPattern>org.apache.flink.avro.registry.glue.shaded.com.amazonaws</shadedPattern>
+                                </relocation>
+                                <relocation>
+                                    <pattern>org.apache.kafka</pattern>
+                                    <shadedPattern>org.apache.flink.avro.registry.glue.shaded.org.apache.kafka</shadedPattern>
+                                </relocation>
+                                <relocation>
+                                    <pattern>org.reactivestreams</pattern>
+                                    <shadedPattern>org.apache.flink.avro.registry.glue.shaded.org.reactivestreams</shadedPattern>
+                                </relocation>
```
They don't seems to be actually used by the AVRO SerDe: excluded

nicusX commented:
@z3d1k I addressed most of the comments. But I am not sure how to implement the IT, in particular for a *format* using an external registry.
Any reference or example I can follow?
Most of formats lack of IT. And this one requires an external system

The only similar format, [flink-avro-confluent-registry](https://github.com/apache/flink/tree/master/flink-formats/flink-avro-confluent-registry/src/test/java/org/apache/flink/formats/avro/registry/confluent), doesn't have any IT.

The [GRS Serialization/DeserializationSchema IT](https://github.com/nicusX/flink-connector-aws/blob/27db878e7da1d56d395ce80984912b16e5b916d4/flink-connector-aws-e2e-tests/flink-formats-avro-glue-schema-registry-e2e-tests/src/test/java/org/apache/flink/glue/schema/registry/test/GlueSchemaRegistryAvroKinesisITCase.java#L189) relies on a [AWS account](https://github.com/nicusX/flink-connector-aws/blob/27db878e7da1d56d395ce80984912b16e5b916d4/flink-connector-aws-e2e-tests/flink-formats-avro-glue-schema-registry-e2e-tests/src/test/java/org/apache/flink/glue/schema/registry/test/GlueSchemaRegistryAvroKinesisITCase.java#L69C61-L69C91) that I assume is available during the CI/CD build. Any reference on how to leverage this?

jherico commented:
Did you guys seriously repurpose my existing PR against the old codebase and not credit me in any way?

fsenart commented on flink-formats-aws/flink-avro-glue-schema-registry/src/main/java/org/apache/flink/formats/avro/glue/schema/registry/AvroGlueFormatOptions.java:
```diff
@@ -0,0 +1,92 @@
+
+/** Options for AWS Glue Schema Registry Avro format. */
+@PublicEvolving
+public class AvroGlueFormatOptions {
```
WDYT about aligning the naming with AWSSchemaRegistryConstants (for all config options below)?
Example:
```
    public static final ConfigOption<String> AWS_REGION =
            ConfigOptions.key(AWSSchemaRegistryConstants.AWS_REGION)
                    .stringType()
                    .noDefaultValue()
                    .withDescription("AWS region");
```
